File: apps/ingestor/internal/event/ua.go
package event

import (
	"strings"

	"github.com/mileusna/useragent"
)

// Device buckets we store.
const (
	DeviceDesktop = "desktop"
	DeviceMobile  = "mobile"
	DeviceTablet  = "tablet"
)

// DeviceInfo is the device data derived from a User-Agent. The UA itself is
// discarded after this.
type DeviceInfo struct {
	Browser        string
	BrowserVersion string
	OS             string
	OSVersion      string
	Device         string
}

func majorVersion(v string) string {
	major, _, _ := strings.Cut(v, ".")
	return major
}

// ParseUserAgent parses a UA string into stored device columns.
func ParseUserAgent(ua string) DeviceInfo {
	r := useragent.Parse(ua)
	device := DeviceDesktop
	switch {
	case r.Tablet:
		device = DeviceTablet
	case r.Mobile:
		device = DeviceMobile
	}
	return DeviceInfo{
		Browser:        r.Name,
		BrowserVersion: majorVersion(r.Version),
		OS:             r.OS,
		OSVersion:      r.OSVersion,
		Device:         device,
	}
}

File: apps/ingestor/src/lib/ua.ts
import { UAParser } from 'ua-parser-js';
import type { DeviceType } from '@seyr/shared';

export interface DeviceInfo {
	browser: string;
	browser_version: string;
	os: string;
	os_version: string;
	device: DeviceType;
}

/** Map ua-parser's device type to our three buckets (desktop is the default). */
function deviceType(type: string | undefined): DeviceType {
	if (type === 'mobile') return 'mobile';
	if (type === 'tablet') return 'tablet';
	return 'desktop';
}

/** Parse a User-Agent string into stored device columns. UA itself is discarded. */
export function parseUserAgent(ua: string): DeviceInfo {
	const r = UAParser(ua);
	return {
		browser: r.browser.name ?? '',
		browser_version: (r.browser.version ?? '').split('.')[0] ?? '', // major only
		os: r.os.name ?? '',
		os_version: r.os.version ?? '',
		device: deviceType(r.device.type)
	};
}
